tests/core/test_utilities.py

```diff
@@ -913,6 +914,81 @@ def test_fit_plane_to_points():
     )
 
 
+DEFAULT_PRINCIPAL_AXES = [[0.0, 0.0, 1.0], [0.0, 1.0, 0.0], [-1.0, 0.0, 0.0]]
+
+CASE_RANK0 = (
+    0,
+    [[0, 0, 0], [0, 0, 0]],
+    [DEFAULT_PRINCIPAL_AXES],
+)
+CASE_RANK1 = (
+    1,
+    [[0, 0, 0], [1, 0, 0]],
+    [[1.0, 0.0, 0.0], [0.0, 0.0, 1.0], [0.0, -1.0, 0.0]],
+)
+
+CASE_RANK2 = (
+    2,
+    [[1, 2, 3], [3, 2, 1]],
+    [
+        [-0.70710678, 0.0, 0.70710678],
+        [0.0, 1.0, 0.0],
+        [-0.70710678, 0.0, -0.70710678],
+    ],
+)
+CASE_RANK3 = (
+    3,
+    pv.examples.load_airplane().points,
+    [
+        [8.1159601e-07, -9.9255711e-01, -1.2178054e-01],
+        [-1.0000000e00, -8.0911440e-07, -6.9828459e-08],
+        [-2.9225653e-08, 1.2178054e-01, -9.9255711e-01],
+    ],
+)
+
+
+@pytest.mark.parametrize(
+    ('rank', 'points', 'expected_axes'),
+    [CASE_RANK0, CASE_RANK1, CASE_RANK2, CASE_RANK3],
+    ids=['rank0', 'rank1', 'rank2', 'rank3'],
+)
+def test_principal_axes(rank, points, expected_axes):
+    assert np.linalg.matrix_rank(points) == rank
+
+    axes = principal_axes(points)
+    assert np.allclose(axes, expected_axes, atol=1e-7)
+
+    assert np.array_equal(np.cross(axes[0], axes[1]), axes[2])
+    assert np.allclose(np.linalg.norm(axes[0]), 1)
+    assert np.allclose(np.linalg.norm(axes[1]), 1)
+    assert np.allclose(np.linalg.norm(axes[2]), 1)
+
+    assert type(axes) is np.ndarray
+
+    _, sizes = principal_axes(points, return_sizes=True)
+    assert sizes[0] >= sizes[1]
+    assert sizes[1] >= sizes[2]
+    assert type(sizes) is np.ndarray
```

Comment:
I recommend also adding a dedicated `return_sizes` test where the input are points from an ellipsoid (shifted away from the origin for good measure). If our code works then the sizes should be `np.allclose()` to the ellipsoid's corresponding parameters.

Reply:
> I recommend also adding a dedicated `return_sizes` test where the input are points from an ellipsoid (shifted away from the origin for good measure). If our code works then the sizes should be `np.allclose()` to the ellipsoid's corresponding parameters.

Done, added in https://github.com/pyvista/pyvista/pull/6346/commits/8264409e45fb6fd3ca4a65bfd7c6122fb6c6500f

Though I used a simplified dataset with 6 points instead of the ellipsoid. I noticed  that `sizes` correlates with the mesh density. So, more points -> larger sizes. Using the ellipsoid at first gave me much larger values than the input xradius, yradius, zradius... and simplifying to 6 points reduced the values of `sizes`

Reply:
I was able to make sense of the returned sizes. You have to convert them to ratios/percentages for it to make sense.

The test was updated in https://github.com/pyvista/pyvista/pull/6346/commits/c9bf952c3d676679cb32ea5ab91f2f759db51566

and the docstring example updated too in https://github.com/pyvista/pyvista/pull/6346/commits/c0664d71f3bf0006b970e6533842fd515fec2c54

Other than formatting or minor edits this should now be resolved.